Argon2 implementation

Jujumba opened the pull request:
- [X] Basic implementation
- [X] Data input
- [X] Salt input
- [ ] rounds/memory/parallelism/secret input
- [X] Variant/Version input
- [x] Bugs 🪲🔍❌
- [x] Hash verification 

closes #4

TheBestTvarynka commented:
Ping me when you'll be ready for the review

Jujumba commented on README.md:
Hope you don't mind on these changes on this branch

TheBestTvarynka replied:
okay

Jujumba commented:
I'm truly sorry, this would take much more time than I expected (due to lack of time and I've accidentally made some weird bugs I can't catch yet 😇 ).

TheBestTvarynka commented:
no problem. I can understand you 😅

Jujumba commented:
I need a bit of your help, @TheBestTvarynka 😅

For some weird reason, the input (and all other elements) do not show up 🤷
![image](https://github.com/TheBestTvarynka/crypto-helper/assets/70100129/4b583bc1-6d48-402d-b981-3d5f98735547)
And after selecting `Argon2` input, this problem spreads to all other inputs

Jujumba commented:
> @Jujumba I'll look at this tonight after the work 🙏 
> 
> thank you

You would better sleep at night 😅
Check this one out when you are truly free.

@TheBestTvarynka

TheBestTvarynka commented:
I finally have time to look at it. first inspection:

![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/6d891207-7493-43ad-be9a-729f7dade42f)

TheBestTvarynka commented:
![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/ef76ff24-f5ca-4831-ad94-48da95f04b36)

@Jujumba, this is the cause of your problem (see the image above)

Jujumba commented:
> ![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/ef76ff24-f5ca-4831-ad94-48da95f04b36)
> 
> 
> 
> @Jujumba, this is the cause of your problem (see the image above).
> 
> I changed it to:
> 
> ```rust
> 
> impl From<&Argon2Action> for bool {
> 
>     fn from(value: &Argon2Action) -> Self {
> 
>         match value {
> 
>             Argon2Action::Hash(_) => false,
> 
>             Argon2Action::Verify(_) => true,
> 
>         }
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> And now it works:
> 
> ![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/cfcd26ad-33e0-4534-b7a9-cc76a983ff8f)
> 
> 

Thanks 🙏

Jujumba commented:
What do you think, should I implement `Argon2HashAction`
```rust
pub struct Argon2HashAction {
    pub memory: u32,
    pub iters: u32,
    pub paralelism: u32,
    pub output_len: u32,
    pub salt: Vec<u8>,
    pub variant: Argon2Variant,
    pub version: Argon2Version
}
```
Or it's better to use "hashing-config" from `rust-argon2` crate?

I tend to the second option, but maybe it's more convenient to you to have home-baked configuration.

@TheBestTvarynka

TheBestTvarynka commented:
@Jujumba you can take the one from the library. I'm OK with it. I don't think we'll need an additional functionality, so go ahead

Jujumba commented:
Nevermind, I just realized it's impossible due to serialization :)

Jujumba commented:
I think computational part is somehow ready for review

@TheBestTvarynka

Jujumba commented:
Ah, and sorry if some lines you didn't expect to be changed were changed

`cargo fmt` did its job

TheBestTvarynka commented:
great, can you fix merge conflicts first?

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
+    pub iters: u32,
+    pub paralelism: u32,
+    pub output_len: u32,
+    pub salt: Vec<u8>,
```
```suggestion
    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
    pub salt: Vec<u8>,
```

TheBestTvarynka replied:
the same for the `secret` and `data`

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
+    type Error = String;
+    fn try_from(hash_action: &'a Argon2HashAction) -> Result<Self, Self::Error> {
+        argon2::password_hash::Salt::from_b64(
+            std::str::from_utf8(&hash_action.salt).map_err(|_| "Non UTF-8 salt".to_owned())?,
```
salt should not be UTF-8. salt is just random bytes.
you should encode them as base64 and then pass them into `Salt::from_b64`

Jujumba replied:
But we need to somehow get a string. Personally, I don't know the way to get a non-utf8 string in rust (besides [`std::from_utf8_lossy`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy), or this is a sound solution? ).

TheBestTvarynka replied:
1. you store salt as `Vec<u8>`
2. when you need to provide it to the `argon2`, then do the following: `Salt::from_b64(base64::encode(&hash_action.salt))`

Jujumba replied:
Fixed 🫡

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
+    }
+}
+
+impl Default for Argon2Action {
+    fn default() -> Self {
+        Self::Hash(Argon2HashAction::default())
+    }
+}
```
this implementation can be derived, as I understand

```rust
#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize, Default)]
pub enum Argon2Action {
    #[default]
    Hash(Argon2HashAction),
    Verify(Vec<u8>),
}
```

Jujumba replied:
No, it can't be derived :(
`#[default]` attribute works only on unit variants

TheBestTvarynka replied:
😢

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
+    }
+}
+
+impl Default for Argon2Input {
+    fn default() -> Self {
+        Self {
+            action: Argon2Action::default(),
+            data: Vec::new(),
+        }
+    }
+}
```
can be derived

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
+
+impl From<&Argon2Action> for bool {
+    fn from(value: &Argon2Action) -> Self {
+        value.into()
```
are you sure this code will not cause a recursion?

Jujumba replied:
Oops, thought I changed it 🤭

TheBestTvarynka commented on src/crypto_helper/info.rs:
```diff
@@ -73,6 +73,9 @@ fn get_algorithm_info(algorithm: &Algorithm) -> Html {
             </span>
         },
+        Algorithm::Argon2(_) => html! {
+            <span>{"Use Argon2 to encrypt/verify your data."}</span>
```
add link to the rfc: https://www.rfc-editor.org/rfc/inline-errata/rfc9106.html

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
this component isn't ready yet, am I right?

Jujumba replied:
> this component isn't ready yet, am I right?

yes, thus I only asked to review "computational" part 😅

TheBestTvarynka reviewed (changes requested):
thank you for your work. I appreciate this. I left some comments. ping me when you fix them

Jujumba commented:
Recently I've cleaned my laptop from windows, but forgot to push some changes before this, so I'll start some things over  😅 😅

@TheBestTvarynka

TheBestTvarynka commented:
no problem ❤️

Jujumba commented:
I'm going to ask a bit stupid question, but I've been banging my head over the wall for the last 20 minutes with that: How to fix these merge conflicts?..

Sorry for the inconvenience 

@TheBestTvarynka

Jujumba commented:
> Recently I've cleaned my laptop from windows, but had forgotten to push some changes, so I'll start some things over. Sorry 😅 😅
> 
> @TheBestTvarynka

Do not worry, as figured out it was some minor changes, I've already rewritten them :)

Jujumba commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+                if !get_hash_verify_switch_state.emit(()) {
+                    html! {
+                        <>
+                            {"Version"}
+                            <select onchange={set_version}>
+                                <option value="Argon10">{"Argon10"}</option>
+                                <option value="Argon13">{"Argon13"}</option>
+                            </select>
```
Could you, please, review this (also see `set_version` closure above) ? 

I highly doubt that this an idiomatic way to get, amount of `.clone()` calls also scares a bit (though it seems to be normal :( ), but I failed to come up with something better :sob: 

@TheBestTvarynka

TheBestTvarynka replied:
I'll check it in around two days. Sorry for such a delay. I'm just very busy now 😞

Jujumba replied:
No problem 🥺

TheBestTvarynka commented:
> How to fix these merge conflicts?..

don't be bothered with it for now. just finish the argon2 and deal with conflicts later

Jujumba commented:
By the way, do you mind adding MRCV (minimum Rust compiler version) to readme?

@TheBestTvarynka

TheBestTvarynka commented:
> do you mind adding MRCV (minimum Rust compiler version) to readme?

the MRCV is the latest one 😅 
I'll use the latest Rust if I want to use any of its features. So, we don't have any specific minimum supported version.

TheBestTvarynka commented:
> I'll use the latest Rust if I want to use any of its features.

maybe we should add a note about it in the readme 😄

TheBestTvarynka commented on src/about.rs:
```diff
@@ -7,7 +7,11 @@ pub fn about() -> Html {
             <span>{"Crypto-helper"}</span>
             <span>{"The crypto-helper is an online app that helps to work with the diferent crypto algorithms:"}</span>
             <ul>
+                <li>{"Argon2"}</li>
+                <li>{"BCrypt"}</li>
                 <li>{"MD5"}</li>
+                <li>{"Argon2"}</li>
+                <li>{"BCRYPT"}</li>
```
Duplicate?

Jujumba replied:
Oops ;)

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,143 @@ pub struct ZlibInput {
+            .p_cost(value.paralelism)
+            .t_cost(value.iters)
+            .build()
+            .unwrap()
```
I know that it'll never panic, but add the message anyway:

```suggestion
            .expect("argon2::ParamsBuilder should never fail")
```

TheBestTvarynka replied:
https://github.com/TheBestTvarynka/crypto-helper/pull/47/files#r1483254713 this one is not fixed

TheBestTvarynka reviewed (changes requested):
good code. I have only two comments

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
I'll review this component more carefully later. It seems good foe now, but I think we can make optimize it

Jujumba replied:
Thanks, let me now when you do 🫡

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+                if !get_hash_verify_switch_state.emit(()) {
+                    html! {
+                        <>
+                            {"Version"}
+                            <select onchange={set_version}>
+                                <option value="Argon10">{"Argon10"}</option>
+                                <option value="Argon13">{"Argon13"}</option>
+                            </select>
```
1. you can remove the "version" label. it's obvious even without it.
2. style this component a little bit by adding a "base-input" CSS class to it.

```suggestion
                            <select onchange={set_version} class="base-input">
                                <option value="Argon10">{"Argon10"}</option>
                                <option value="Argon13">{"Argon13"}</option>
                            </select>
```

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
+            {
+                if !get_hash_verify_switch_state.emit(()) {
```
you can simplify it and remove the callback:

```suggestion
                if let Argon2Action::Hash(_) = &props.input.action {
```

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+
+    let input_setter = props.argon2_input_setter.clone();
+    let input = props.input.clone();
+    let set_version = move |event: web_sys::Event| {
+        let target = event.target().expect("NO TARGET");
+        let version: String = js_sys::Reflect::get(target.as_ref(), &"value".into())
+            .unwrap()
+            .as_string()
+            .unwrap();
+
+        let input = input.clone();
+        let input = input.set_version(version.as_str().try_into().unwrap());
+        log::debug!("{input:?}");
+        input_setter.emit(input.clone());
+    };
```
this callback can be simplified:

```suggestion
    let set_version = move |event: yew::html::onchange::Event| {
        let html_element: HtmlInputElement = event.target_unchecked_into();
        let input = input.set_version(html_element.value().as_str().try_into().unwrap());
        input_setter.emit(input.clone());
    };

```

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
+            {
```
actually, you can replace this block of the code with:

```rust
            {if let Argon2Action::Hash(_) = &props.input.action {html! {
                <select onchange={set_version} class="base-input">
                    <option value="Argon10">{"Argon10"}</option>
                    <option value="Argon13">{"Argon13"}</option>
                </select>
            }} else {html! {}}}
```

:smiley:

Jujumba replied:
Replace
```rust
  </div>
            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}

            {
```
with
```rust
   {if let Argon2Action::Hash(_) = &props.input.action {html! {
                <select onchange={set_version} class="base-input">
                    <option value="Argon10">{"Argon10"}</option>
                    <option value="Argon13">{"Argon13"}</option>
                </select>
            }} else {html! {}}}
```
? 😵‍💫

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,89 @@
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
```
can you also all the byte input for the salt? currently, I can not enter the salt and get this error every time:

![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/7e6304e1-fccc-42ee-ab68-10477a0bc1a1)

TheBestTvarynka commented on src/crypto_helper/computations.rs:
```diff
@@ -92,3 +94,38 @@ pub fn process_zlib(input: &ZlibInput) -> Result<Vec<u8>, String> {
+                hash_action.version.into(),
+                hash_action.into(),
+            );
+            let b64 = base64::encode(&hash_action.salt);
+            let salt = argon2::password_hash::Salt::from_b64(&b64).unwrap();
```
1. add the salt generation of the `hash_action.salt` is empty
2. handle errors here properly. too short salt will produce an error here

```suggestion
            let b64 = base64::encode(&hash_action.salt); // handle an empty salt here
            let salt = argon2::password_hash::Salt::from_b64(&b64)
                .map_err(|err| format!("Error while building a salt: {:?}", err))?;
```

TheBestTvarynka reviewed (changes requested):
good code and nice progress. I like your imlementation

Jujumba commented:
You can look on new changes

@TheBestTvarynka 

But there are still bugs (😇)

Like this one:
![image](https://github.com/TheBestTvarynka/crypto-helper/assets/70100129/ffbbeb0a-861e-4e42-a7d4-a68a6712f3e1)

And this one (which seems extra weird)
![image](https://github.com/TheBestTvarynka/crypto-helper/assets/70100129/0181b557-214c-44b5-a982-01d9d04ea3d7)

Or perhaps this is not even a bug 🤷‍♂️

TheBestTvarynka commented:
looks good. about bugs:

> the last one persists despite of salt's length. Any salt ending with = considered invalid

maybe. you should play with base64 formats (with and without padding)

> I've dug a bit deeper: The returned hash is encoded in [b64 with no padding](https://docs.rs/password-hash/latest/password_hash/enum.Encoding.html#variant.B64). Data encoded in which format does function process_argon2 should return?

in bytes. if the output from the lib is in the b64 format, then decode it and return as plain bytes

Jujumba commented:
> maybe. you should play with base64 formats (with and without padding)

Thanks, I've got that working. But the seems to weirdness come the `argon2` crate itself?....

Encoding `Some data` with `Some salt` salt returns this:
`oenotYnYadelI+Fqhs92l9hnk0xyI6+sIeSC7+6OzqM`

"Translated" to UTF-8 it means the following: `赉iץ#jvgLr#!Σ`, what's not a PHC string :(

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
+    }
+}
+impl<'a> TryFrom<&'a str> for Argon2Version {
+    type Error = (); // todo: proper error type
```
Do you still plan to fix it?

Jujumba replied:
Yes

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
+            _ => Err(()),
+        }
+    }
+}
+impl From<Argon2Variant> for argon2::Algorithm {
```
add an empty line between impls

```suggestion
}

impl From<Argon2Variant> for argon2::Algorithm {
```

TheBestTvarynka commented on src/crypto_helper/algorithm.rs:
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
+    }
+}
+impl TryFrom<&str> for Argon2Variant {
+    type Error = ();
```
use proper error type. `String`, for example

TheBestTvarynka commented on src/crypto_helper/computations.rs:
```diff
@@ -92,3 +94,41 @@ pub fn process_zlib(input: &ZlibInput) -> Result<Vec<u8>, String> {
+
+            Ok(hash.into_bytes())
+        }
+        Argon2Action::Verify(data) => {
+            let b64 = base64::encode_config(data, config);
+
+            let hash = argon2::PasswordHash::parse(&b64, argon2::password_hash::Encoding::B64)
+                .map_err(|e| e.to_string())?;
+
+            if argon2::Argon2::default().verify_password(data, &hash).is_ok() {
+                Ok(vec![1])
+            } else {
+                Ok(vec![0])
+            }
+        }
```
The password verification algorithm is wrong. To verify the password, you need two items: the password itself and the argon2 hash of this password

Jujumba replied:
Yeah, it's not ready yet :(

TheBestTvarynka replied:
Password hashing looks good to me. so, password verification is the only thing left

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,103 @@
+                <Switch id={"switch".to_string()} setter={on_hash_verify_switch} state={hash_verify_switch_state}/>
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
```
add a hint for the user

```suggestion
            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("password".into()))}
```

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,103 @@
+                        <option value="Argon2id">{"Argon2id"}</option>
+                    </select>
+                </>
+            }} else {html! {}}}
```
this `if` branch misses the password hash input

TheBestTvarynka reviewed (changes requested):
good code but I always find something to fix :smile: 

1. run `cargo clippy --all-targets --all-features -- -D warnings`
2. run `cargo +nightly fmt --all`

TheBestTvarynka reviewed (approved):
the UI of the argon2 input can be improved, but it's not a blocker

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,120 @@
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("password".into()))}
+            {match &props.input.action {
+                Argon2Action::Hash(hash_action) => html! {
+                    <>
+                        {build_byte_input(hash_action.salt.clone(), salt_setter, Some(BytesFormat::Ascii), Some("salt".into()))}
+                        <select onchange={set_version} class="base-input">
+                            <option value="Argon10">{"Argon10"}</option>
+                            <option value="Argon13">{"Argon13"}</option>
+                        </select>
+                        <select onchange={set_variant} class="base-input">
+                            <option value="Argon2i">{"Argon2i"}</option>
+                            <option value="Argon2d">{"Argon2d"}</option>
+                            <option value="Argon2id">{"Argon2id"}</option>
+                        </select>
+                    </>
+                },
+                Argon2Action::Verify(hash) => html! {
+                    <>
+                        {build_byte_input(hash.clone(), hash_setter, Some(BytesFormat::Ascii), Some("hash".into()))}
+                    </>
+                }}}
```
can be improved:

```suggestion
                Argon2Action::Hash(hash_action) => html! {
                    <div class="horizontal">
                        {build_byte_input(hash_action.salt.clone(), salt_setter, Some(BytesFormat::Ascii), Some("salt".into()))}
                        <div class="vertical">
                            <select onchange={set_version} class="base-input">
                                <option value="Argon10">{"Argon10"}</option>
                                <option value="Argon13">{"Argon13"}</option>
                            </select>
                            <select onchange={set_variant} class="base-input">
                                <option value="Argon2i">{"Argon2i"}</option>
                                <option value="Argon2d">{"Argon2d"}</option>
                                <option value="Argon2id">{"Argon2id"}</option>
                            </select>
                        </div>
                    </div>
                },
                Argon2Action::Verify(hash) => build_byte_input(hash.clone(), hash_setter, Some(BytesFormat::Ascii), Some("hash".into()))
```

TheBestTvarynka commented on src/crypto_helper/input/argon2.rs:
```diff
@@ -0,0 +1,120 @@
+                Argon2Action::Hash(hash_action) => html! {
+                    <>
+                        {build_byte_input(hash_action.salt.clone(), salt_setter, Some(BytesFormat::Ascii), Some("salt".into()))}
+                        <select onchange={set_version} class="base-input">
+                            <option value="Argon10">{"Argon10"}</option>
+                            <option value="Argon13">{"Argon13"}</option>
+                        </select>
+                        <select onchange={set_variant} class="base-input">
+                            <option value="Argon2i">{"Argon2i"}</option>
+                            <option value="Argon2d">{"Argon2d"}</option>
+                            <option value="Argon2id">{"Argon2id"}</option>
+                        </select>
```
I spot a problem here:
![image](https://github.com/TheBestTvarynka/crypto-helper/assets/43034350/d244bfc4-828d-4546-8ab3-ce0862d6505d)
initial values of these `select` inputs are invalid. you don't have _argon1_ version. only _10_ and _13_. the same with argon2 variant

TheBestTvarynka reviewed (changes requested):
1. run `cargo +nightly fmt --all -- --check`
2. run `cargo clippy -- -D warnings`

Jujumba commented:
> 2\. cargo clippy -- -D warnings
I guess I'm not meant to fix that..?
```sh
error: unused imports: `ByteInputProps`, `ByteInput`
 --> src/common/mod.rs:7:40
  |
7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
  |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: unused imports: `BytesViewerProps`, `BytesViewer`
 --> src/common/mod.rs:8:24
  |
8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
  |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^

error: unused import: `CheckboxProps`
 --> src/common/mod.rs:9:30
  |
9 | pub use checkbox::{Checkbox, CheckboxProps};
  |                              ^^^^^^^^^^^^^

error: unused import: `SwitchProps`
  --> src/common/mod.rs:11:26
   |
11 | pub use switch::{Switch, SwitchProps};
   |                          ^^^^^^^^^^^

error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
 --> src/crypto_helper.rs:9:32
  |
9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
  |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^

error: unused import: `jwt::Jwt as JwtData`
  --> src/jwt.rs:11:9
   |
11 | pub use jwt::Jwt as JwtData;
   |         ^^^^^^^^^^^^^^^^^^^
```

Jujumba commented:
> > 2. cargo clippy -- -D warnings
> > I guess I'm not meant to fix that..?
> 
> ```shell
> error: unused imports: `ByteInputProps`, `ByteInput`
>  --> src/common/mod.rs:7:40
>   |
> 7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
>   |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
>   |
>   = note: `-D unused-imports` implied by `-D warnings`
>   = help: to override `-D warnings` add `#[allow(unused_imports)]`
> 
> error: unused imports: `BytesViewerProps`, `BytesViewer`
>  --> src/common/mod.rs:8:24
>   |
> 8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
>   |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^
> 
> error: unused import: `CheckboxProps`
>  --> src/common/mod.rs:9:30
>   |
> 9 | pub use checkbox::{Checkbox, CheckboxProps};
>   |                              ^^^^^^^^^^^^^
> 
> error: unused import: `SwitchProps`
>   --> src/common/mod.rs:11:26
>    |
> 11 | pub use switch::{Switch, SwitchProps};
>    |                          ^^^^^^^^^^^
> 
> error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
>  --> src/crypto_helper.rs:9:32
>   |
> 9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
>   |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^
> 
> error: unused import: `jwt::Jwt as JwtData`
>   --> src/jwt.rs:11:9
>    |
> 11 | pub use jwt::Jwt as JwtData;
>    |         ^^^^^^^^^^^^^^^^^^^
> ```



> 1. run `cargo +nightly fmt --all -- --check`
> 
>     2. run `cargo clippy -- -D warnings`

I guess I'm not meant to fix that, as there no lints applied to my code (?)

```shell
error: unused imports: `ByteInputProps`, `ByteInput`
 --> src/common/mod.rs:7:40
  |
7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
  |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: unused imports: `BytesViewerProps`, `BytesViewer`
 --> src/common/mod.rs:8:24
  |
8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
  |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^

error: unused import: `CheckboxProps`
 --> src/common/mod.rs:9:30
  |
9 | pub use checkbox::{Checkbox, CheckboxProps};
  |                              ^^^^^^^^^^^^^

error: unused import: `SwitchProps`
  --> src/common/mod.rs:11:26
   |
11 | pub use switch::{Switch, SwitchProps};
   |                          ^^^^^^^^^^^

error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
 --> src/crypto_helper.rs:9:32
  |
9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
  |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^

error: unused import: `jwt::Jwt as JwtData`
  --> src/jwt.rs:11:9
   |
11 | pub use jwt::Jwt as JwtData;
   |         ^^^^^^^^^^^^^^^^^^^
```
```

TheBestTvarynka commented:
> I guess I'm not meant to fix that?

@Jujumba please, fix the formatting ones but leave the `clippy` errors

TheBestTvarynka commented:
@Jujumba will you fix the merge conflicts or do I have to do it?

Jujumba commented:
> > I guess I'm not meant to fix that?
> 
> 
> 
> @Jujumba please, fix the formatting ones but leave the `clippy` errors

They are already fixed, aren't they?

Jujumba commented:
> @Jujumba will you fix the merge conflicts or do I have to do it?

I _really_ would like to delegate that work on you. Sorry 🙁
